### ISSUE TITLE
Can we change Reducer from protocol into function type?

### ISSUE THREAD

**Issue by soapsign** [ReSwift/ReSwift#118]
Most of the time we need to create a struct to wrap the underlying `handleAction`. If the same functionality can be done with typealias. Should we change to typealias so we don't need to wrap the real  `handleAction` inside a struct.

**Comment by Ben-G** [comment]
Yes, we could potentially do this! The reducer protocol is an artifact from an older ReSwift version. One of the advantages of using a type instead of a free function is that the error message around type mismatches in the store and the root-reducer is slightly better.

However, our docs have been suggesting using free functions for any but the root reducer function; so it might be more consistent to change this.

I don't have a strong opinion either way. I would love to see the change on a branch to see the impact it would have.

@ReSwift/reswift any opinions on this?

**Comment by DivineDominion** [comment]
A reducer type doesn't have much value in my app TableFlip so far compared to a free function.

**Comment by hartbit** [comment]
I also agree that they should be free functions 👍

**Comment by Ben-G** [comment]
I believe @agentk was in favor of this as well. If that's the case I'd be happy to deprecate the `Reducer` protocol with the next release and add support for free functions.

**Comment by Qata** [comment]
I like the idea of deprecating `Reducer`, but what I like even more is turning `Reducer` into a concrete struct that allows mapping and concatenating `Reducer`s with ease. I've implemented this [in ReactiveReSwift](https://github.com/ReSwift/ReactiveReSwift/blob/master/ReactiveReSwift/CoreTypes/Reducer.swift) and I'm very happy with the results. Here's a reimplementation of the GitHub Example code using this new structure.

```swift
let appReducer = Reducer(authenticationReducer, bookmarksReducer, repositoriesReducer)

let authenticationReducer = Reducer<State> { action, state in
    var state = state

    switch action {
    case let action as SetOAuthURL:
        state.authenticationState.oAuthURL = action.oAuthUrl
    case let action as UpdateLoggedInState:
        state.authenticationState.loggedInState = action.loggedInState
    default:
        break
    }

    return state
}

let bookmarksReducer = Reducer<State> { action, state in
    var state = state

    if let action = action as? CreateBookmark {
        let bookmark = (route: action.route, routeSpecificData: action.routeSpecificData)
        state.bookmarks.append(bookmark)
    }

    return state
}

let repositoriesReducer = Reducer<State> { action, state in
    var state = state

    if let action = action as? SetRepostories {
        state.repositories = action.repositories
    }

    return state
}```

**Comment by hartbit** [comment]
This looks very nice 👍🏻

**Comment by DivineDominion** [comment]
Love it!

**Comment by Ben-G** [comment]
@Qata agree that this looks nice! One thought: using only functions it's easy to compose reducers that operate on a substate of the overall state. E.g. in the GitHub example we only pass the bookmark state to the bookmark reducer. This enforces separation of concerns within reducers. Do you have an idea if/how this would be possible with your approach?

**Comment by Ben-G** [comment]
Finally opened #177, that serves the original request of making Reducer a function type. While I'm a big fan of @Qata suggestion of revamping the middleware API (#175). I think that having the reducer being a function type has advantages for composition that I mentioned above.

You can see an example here: https://github.com/ReSwift/GitHubBrowserExample/blob/master/SwiftFlowGitHubBrowser/Reducers/AppReducer.swift#L16-L19

**Comment by Ben-G** [comment]
Shipped with #177.
